File: 30 Days of Code/Day 15 - Linked Lists/Solved.py
# Define a Node class to represent each element in the linked list
class Node:
    def __init__(self, data):
        self.data = data  # Store the data value of the node
        self.next = None  # Initialize the next pointer to None

# Define a Solution class to handle linked list operations
class Solution: 
    # Method to display the elements of the linked list
    def display(self, head):
        current = head  # Start from the head of the linked list
        while current:  # Traverse the list until the end
            print(current.data, end=' ')  # Print the data of the current node
            current = current.next  # Move to the next node

    # Method to insert a new node at the end of the linked list
    def insert(self, head, data): 
        # Create a new node with the given data
        new_node = Node(data)
        # If the linked list is empty, return the new node as the head
        if head is None:
            return new_node
        else:
            # Traverse the list to find the last node
            current = head
            while current.next:
                current = current.next
            # Link the last node to the new node
            current.next = new_node
        # Return the head of the updated linked list
        return head
        
    # Method to display the elements of the linked list (duplicate definition, can be removed)
    def display(self, head):
        current = head  # Start from the head of the linked list
        while current:  # Traverse the list until the end
            print(current.data, end=' ')  # Print the data of the current node
            current = current.next  # Move to the next node

# Create an instance of the Solution class
mylist = Solution()

# Read the number of elements to be inserted into the linked list
T = int(input())

# Initialize the head of the linked list to None
head = None

# Loop to read data and insert nodes into the linked list
for i in range(T):
    data = int(input())  # Read the data for the new node
    head = mylist.insert(head, data)  # Insert the new node into the linked list

# Display the elements of the linked list
mylist.display(head)

File: 30 Days of Code/Day 15 - Linked List/Solved.c
#include <stdlib.h>
#include <stdio.h>	

// Define a structure for a Node in the linked list
typedef struct Node
{
    int data;              // Data stored in the node
    struct Node* next;     // Pointer to the next node in the list
}Node;

// Function to insert a new node at the end of the linked list
Node* insert(Node *head, int data)
{
    // Allocate memory for a new node and initialize its data and next pointer
    Node* newNode = (Node*)malloc(sizeof(Node));
    newNode->data = data;
    newNode->next = NULL;
    
    // If the list is empty, the new node becomes the head
    if(head == NULL){
        return newNode;
    } else {
        // Traverse the list to find the last node
        Node* temp = head;
        while (temp->next != NULL) {
            temp = temp->next;
        }
        // Link the new node to the end of the list
        temp->next = newNode;
    }
    // Return the head of the list
    return head;
}

// Function to display the elements of the linked list
void display(Node *head)
{
    Node *start = head;  // Start from the head of the list
    while(start)         // Traverse the list until the end
    {
        printf("%d ", start->data);  // Print the data of the current node
        start = start->next;         // Move to the next node
    }
}

int main()
{
    int T, data;  // T: number of elements to insert, data: value of each element
    scanf("%d", &T);  // Read the number of elements
    Node *head = NULL;  // Initialize the head of the list as NULL
    
    // Loop to insert T elements into the linked list
    while(T-- > 0){
        scanf("%d", &data);  // Read the data for the new node
        head = insert(head, data);  // Insert the new node into the list
    }
    display(head);  // Display the elements of the linked list
}
